### PR TITLE
Improved generation of Pipeline features for extern functions

### PR DESCRIPTION
**Pull request by benoitsteiner**
Leverage the proxy (if any) to generate pipeline features for extern functions. There's still work to do to model the memory accesses, but we can already have a sense of how much computation is done by the black box (assuming that the proxy is accurate), which should help the auto-scheduler further down the road. Note that this PR doesn't address the generation of schedule features, which will be handled in a follow up PR.

### CODE REVIEW

File: apps/autoscheduler/FunctionDAG.cpp [hunk 1]
```diff
@@ -996,66 +996,83 @@ void FunctionDAG::featurize() {
...
-            for (auto v : def.values()) {
-                featurizer.visit_store_args(node.func.name(), v.type(), def.args());
-                v = common_subexpression_elimination(simplify(v));  // Get things into canonical form
-                v.accept(&featurizer);
-            }
-            for (auto v : def.args()) {
-                v = common_subexpression_elimination(simplify(v));  // Get things into canonical form
+            if (node.func.extern_definition_proxy_expr().get()) {
+                // Extern function call with a proxy implementation specified: generate the featurization from the proxy
+                Expr v = common_subexpression_elimination(simplify(node.func.extern_definition_proxy_expr()));  // Get things into canonical form
```
Comment (steven-johnson) [inline apps/autoscheduler/FunctionDAG.cpp]: Nit: this line seems awfully long.

File: apps/autoscheduler/Makefile [hunk 2]
```diff
@@ -38,6 +38,10 @@ $(BIN)/test_perfect_hash_map: test_perfect_hash_map.cpp PerfectHashMap.h
 	@mkdir -p $(@D)
 	$(CXX) $(CXXFLAGS) $< -o $@
 
+$(BIN)/test_function_dag: test_function_dag.cpp FunctionDAG.h FunctionDAG.cpp ASLog.h ASLog.cpp
```
Comment (steven-johnson) [inline apps/autoscheduler/Makefile]: don't forget to update the CMakeLists.txt too

File: apps/autoscheduler/test_function_dag.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
+#include "FunctionDAG.h"
+#include "Halide.h"
+#include <sstream>
+
+using namespace Halide;
+
+extern "C" int generateOutputVals(
```
Comment (steven-johnson) [inline apps/autoscheduler/test_function_dag.cpp]: generate_output_vals

Comment (benoitsteiner) [reply apps/autoscheduler/test_function_dag.cpp]: fixed.

File: apps/autoscheduler/FunctionDAG.cpp (hunk 1 above)
Comment (benoitsteiner) [reply apps/autoscheduler/FunctionDAG.cpp]: For some reason clang-format won't improve the formatting. I manually split the code.

File: apps/autoscheduler/Makefile (hunk 2 above)
Comment (benoitsteiner) [reply apps/autoscheduler/Makefile]: Done

File: apps/autoscheduler/CMakeLists.txt [hunk 4]
```diff
@@ -153,5 +153,14 @@ endif()
 
 add_executable(test_perfect_hash_map test_perfect_hash_map.cpp)
 
+add_executable(test_function_dag test_function_dag.cpp FunctionDAG.h FunctionDAG.cpp ASLog.h ASLog.cpp)
+target_include_directories(
+  test_function_dag
+  PRIVATE "${HALIDE_INCLUDE_DIR}" "${HALIDE_TOOLS_DIR}")
+target_link_libraries(test_function_dag PRIVATE Halide)
+#if (NOT WIN32)
```
Comment (steven-johnson) [inline apps/autoscheduler/CMakeLists.txt]: Either remove entirely or add comment about why it should stay commented out

**Comment by steven-johnson** [comment]
In the future, may I suggest doing the clang-format step as a separate PR; it makes the meat of the real changes harder to review

File: apps/autoscheduler/CMakeLists.txt (hunk 4 above)
Comment (benoitsteiner) [reply apps/autoscheduler/CMakeLists.txt]: I deleted the code.

**Comment by benoitsteiner** [comment]
Understood. I'll refrain from using clang-format on code that doesn't already follow the formatting standard.

**Comment by steven-johnson** [comment]
> Understood. I'll refrain from using clang-format on code that doesn't already follow the formatting standard.

Oh, please fix the formatting, just consider opening up a *separate* PR to do that with no other changes :-)

(I'm surprised we still have files that haven't been thru the formatter, but perhaps the giant formatting party didn't include apps/?)

File: apps/autoscheduler/Makefile [hunk 5]
```diff
@@ -38,6 +38,10 @@ $(BIN)/test_perfect_hash_map: test_perfect_hash_map.cpp PerfectHashMap.h
...
+	@mkdir -p $(@D)
+	$(CXX) $(CXXFLAGS) $(USE_EXPORT_DYNAMIC) $^ -o $@ $(LDFLAGS) $(LIB_HALIDE) $(HALIDE_SYSTEM_LIBS)
```
Comment (steven-johnson) [inline apps/autoscheduler/Makefile]: This should be `$(filter-out %.h,$^)` instead of `$^`

Comment (benoitsteiner) [reply apps/autoscheduler/Makefile]: Thanks for the tip, done

**Pull request merged**